Userland/Libraries/LibIMAP/Parser.cpp

```diff
@@ -0,0 +1,836 @@
+            auto actual_type = parse_atom();
+            consume(" ");
+            if (actual_type.matches("UIDNEXT")) {
+                auto n = parse_number();
+                m_response.data().set_uid_next(n);
+            } else if (actual_type.matches("UIDVALIDITY")) {
+                auto n = parse_number();
+                m_response.data().set_uid_validity(n);
+            } else if (actual_type.matches("UNSEEN")) {
+                auto n = parse_number();
+                m_response.data().set_unseen(n);
+            } else if (actual_type.matches("PERMANENTFLAGS")) {
+                auto flags = parse_list(+[](StringView x) { return String(x); });
+                m_response.data().set_permanent_flags(move(flags));
+            } else {
+                dbgln("Unknown: {}", actual_type);
+                parse_while(+[](u8 x) { return x != ']'; });
+            }
+            consume("]");
+            parse_while(+[](u8 x) { return x != '\r'; });
+            consume("\r\n");
+        } else {
+            parse_while(+[](u8 x) { return x != '\r'; });
+            consume("\r\n");
+        }
+    } else if (match("SEARCH")) {
+        Vector<unsigned> ids;
+        while (!match("\r\n")) {
+            consume(" ");
+            auto id = parse_number();
+            ids.append(id);
+        }
+        m_response.data().set_search_results(move(ids));
+    } else if (match("BYE")) {
+        auto message = parse_while(+[](u8 x) { return x != '\r'; });
+        consume("\r\n");
+        m_response.data().set_bye(message.is_empty() ? Optional<String>() : Optional<String>(message));
+    } else if (match("STATUS")) {
+        consume(" ");
+        auto mailbox = parse_astring();
+        consume(" (");
+        auto status_item = StatusItem();
+        status_item.set_mailbox(mailbox);
+        while (!match(")")) {
+            auto status_att = parse_atom();
+            consume(" ");
+            auto value = parse_number();
+
+            auto type = StatusItemType::Recent;
+            if (status_att.matches("MESSAGES")) {
+                type = StatusItemType::Messages;
+            } else if (status_att.matches("UNSEEN")) {
+                type = StatusItemType::Unseen;
+            } else if (status_att.matches("UIDNEXT")) {
+                type = StatusItemType::UIDNext;
+            } else if (status_att.matches("UIDVALIDITY")) {
+                type = StatusItemType::UIDValidity;
+            } else if (status_att.matches("RECENT")) {
+                type = StatusItemType::Recent;
+            } else {
+                dbgln("Unmatched status attribute: {}", status_att);
+                m_parsing_failed = true;
+            }
+
+            status_item.set(type, value);
+
+            if (!end() && m_buffer[position] != ')')
+                consume(" ");
+        }
+        m_response.data().set_status(move(status_item));
+        match(" "); // not in the spec but the Outlook server sends a space, idk why?
```

Comment:
Based one prior info on lexer terminology, `match` won't consume anything
Maybe this could be a `consume` overload if it does?

Reply:
Uh.. well I'm kinda a noob to parsing , but I thought match tries to consume a character and returns whether or not it was successful, whereas consume always consumes a character and errors out if it fails? 

Like for example, this is the [Wren language compiler](https://github.com/wren-lang/wren/blob/3c43de7485dd49b1c2a79904ea3d7c79ffc81cdf/src/vm/wren_compiler.c#L661-L667):

```c 
// If the current character is [c], consumes it and returns `true`.
static bool matchChar(Parser* parser, char c)
{
  if (peekChar(parser) != c) return false;
  nextChar(parser);
  return true;
}
```

So I dunno? I can rename `match` to `try_consume` or something if you'd prefer though.

Reply:
I guess different people use it in different ways...
Yeah, `try_consume` sounds okay